### PR TITLE
Decode HTML Reserved Characters (>,<,&,").

### PR DESCRIPTION
**Pull request by samouri**
Followup to https://github.com/ampproject/worker-dom/pull/947. I think it may make sense to decode these 4 in particular, given their prominence in html encoding. See https://developer.mozilla.org/en-US/docs/Glossary/Entity

### CODE REVIEW

File: src/third_party/html-parser/html-parser.ts [hunk 1]
```diff
@@ -232,3 +232,31 @@ export function parse(data: string, rootElement: Element) {
...
+
+/**
+ * Decodes HTML Entities.
+ * Currently only works for numeric entities, as well as the 4 named entities
+ * required for encoding HTML: &, ", <, >.
+ * See https://developer.mozilla.org/en-US/docs/Glossary/Entity.
+ *
+ * TODO: create solution for other named entities.
+ */
+const RESERVED_CHARACTERS: { [key: string]: string } = { amp: '&', lt: '<', gt: '>', quot: '"' };
```
Comment (jridgewell) [inline src/third_party/html-parser/html-parser.ts:244]: Since we’re doing unchecked property lookups, make this a prototype-less object.

Comment (samouri) [reply src/third_party/html-parser/html-parser.ts:244]: How? is that the `Object.create(null)` trick?

**Review by kristoferbaxter — APPROVED** [review]

File: src/third_party/html-parser/html-parser.ts (hunk 1 above)
Comment (jridgewell) [reply src/third_party/html-parser/html-parser.ts:244]: Yah. Or, `{ __proto__: null, amp: '&', … }` which is a literal syntax equivalent.

File: src/third_party/html-parser/html-parser.ts [hunk 2]
```diff
@@ -233,15 +233,30 @@ export function parse(data: string, rootElement: Element) {
...
 
-    // 1114111 is the largest valid unicode codepoint.
-    if (isNaN(code) || code > 1114111) {
-      return s;
+      return String.fromCodePoint(code) || s;
     }
 
-    return String.fromCodePoint(code) || s;
+    // Named entity. Only HTML reserved entities are currently supported.
+    return RESERVED_CHARACTERS[namedEntity] || s;
```
Comment (jridgewell) [inline src/third_party/html-parser/html-parser.ts:260]: Oh, I think we need to lowercase, too?

File: src/third_party/html-parser/html-parser.ts (hunk 1 above)
Comment (samouri) [reply src/third_party/html-parser/html-parser.ts:244]: can't do that with my current typing :/

File: src/third_party/html-parser/html-parser.ts (hunk 2 above)
Comment (samouri) [reply src/third_party/html-parser/html-parser.ts]: yeah, good call. Done

File: src/third_party/html-parser/html-parser.ts (hunk 1 above)
Comment (samouri) [reply src/third_party/html-parser/html-parser.ts]: updated the typing to allow for the null value. done

**Review by jridgewell — APPROVED** [review]

**Pull request merged**